compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java

```diff
@@ -487,58 +496,64 @@ public void visit(BLangVariableDef varDefNode) {
     public void visit(BLangAssignment assignNode) {
         assignNode.expr.accept(this);
         BLangExpression varRefExpr = assignNode.varRef;
-        boolean varTaintedStatus = getObservedTaintedStatus();
-        visitAssignment(varRefExpr, varTaintedStatus, assignNode.pos);
+        visitAssignment(varRefExpr, this.taintedStatus, assignNode.pos);
     }
 
     @Override
     public void visit(BLangPostIncrement postIncrement) {
         BLangExpression varRefExpr = postIncrement.varRef;
         varRefExpr.accept(this);
-        boolean varTaintedStatus = getObservedTaintedStatus();
-        visitAssignment(varRefExpr, varTaintedStatus, postIncrement.pos);
+        visitAssignment(varRefExpr, this.taintedStatus, postIncrement.pos);
     }
 
     @Override
     public void visit(BLangCompoundAssignment compoundAssignment) {
         compoundAssignment.varRef.accept(this);
-        boolean varRefTaintedStatus = getObservedTaintedStatus();
+        Boolean varRefTaintedStatus = this.taintedStatus;
+
         compoundAssignment.expr.accept(this);
-        boolean exprTaintedStatus = getObservedTaintedStatus();
-        boolean combinedTaintedStatus = varRefTaintedStatus || exprTaintedStatus;
+        Boolean exprTaintedStatus = this.taintedStatus;
+
+        Boolean combinedTaintedStatus = null;
+        if (varRefTaintedStatus != null && exprTaintedStatus != null) {
+            combinedTaintedStatus = varRefTaintedStatus || exprTaintedStatus;
+        }
         visitAssignment(compoundAssignment.varRef, combinedTaintedStatus, compoundAssignment.pos);
     }
 
-    private void visitAssignment(BLangExpression varRefExpr, boolean varTaintedStatus, DiagnosticPos pos) {
-        // Generate error if a global variable has been assigned with a tainted value.
-        if (varTaintedStatus && varRefExpr instanceof BLangVariableReference) {
-            BLangVariableReference varRef = (BLangVariableReference) varRefExpr;
-            if (varRef.symbol != null && varRef.symbol.owner != null) {
-                if (varRef.symbol.owner instanceof BPackageSymbol || SymbolKind.SERVICE.equals(varRef.symbol.owner.kind)
-                        || SymbolKind.CONNECTOR.equals(varRef.symbol.owner.kind)) {
-                    addTaintError(pos, varRef.symbol.name.value,
-                            DiagnosticCode.TAINTED_VALUE_PASSED_TO_GLOBAL_VARIABLE);
-                    return;
+    private void visitAssignment(BLangExpression varRefExpr, Boolean varTaintedStatus, DiagnosticPos pos) {
+        if (varTaintedStatus != null) {
+            // Generate error if a global variable has been assigned with a tainted value.
+            if (varTaintedStatus && varRefExpr instanceof BLangVariableReference) {
+                BLangVariableReference varRef = (BLangVariableReference) varRefExpr;
+                if (varRef.symbol != null && varRef.symbol.owner != null) {
+                    if (varRef.symbol.owner instanceof BPackageSymbol
```

Comment:
Shall we combine these two if conditions?

Reply:
Yes. Will update.